rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,140 @@
+# UI component model
+
+Authors: Gordon Brander
+
+## Overview
+
+Context: converge on a default UI component model for LLM-generated UI.
+
+## Goals
+
+### Product goals
+
+- Easy for LLMs to generate
+    - Leverages patterns and/or frameworks that are widely present in the training data, or can be learned within a small context window.
+- Easy for humans to edit
+    - [Maybe it’s not even code?](https://x.com/threepointone/status/1792930000766677034) Or maybe it’s a hybrid of code and plain language?
+- Leverages familiar or established patterns for UI development
+- Conformable with existing web FE toolchains.
+
+### Technical goals
+
+- Components are **encapsulated** (P1)
+    - A component may control its child tree, but not its siblings or parent
+- Components have **inputs**, and **output UI** and **events** (P1)
+    - Components can be understood as pure-ish functions that receive props and return a view description (we may allow cheats for local component state ala hooks)
+    - Components are **black boxes**
+    - Components are decoupled and only communicate via input and output channels.
+- Component **inputs** and **outputs** are **statically-typed** (P1)
+    - E.g. via TypeScript
+    - Allows the runtime to enforce data policies on component
+- Components have **local state** (P1)
+    - State is encapsulated within component
+    - Components may pass state down to child components as input
+        - E.g. the React **[“lifting state up” pattern](https://legacy.reactjs.org/docs/lifting-state-up.html)**.
+    - Local state may be **persisted**.
+        - If it isn’t, it is **ephemeral**, and lasts for the lifetime of the component.
```

Comment:
I've been wondering about the semantics around "local" here.

If you have a Module named "Dropdown" expressing a UI component you have a few options for handling the "open" state.

1. totally encapsulated: no-one but the component knows this state exists (standard React)
2. wrapped & exported: state is held in an inner Behaviour but is exported as a readonly Event
3. inversion-of-control: UI components only declare their state, they do not _own_ it

If we're going to implement persisted internal state it seems like we need (3) regardless of the skinning layer that goes on top? Is there a benefit to "hiding" state vs. simply ignoring it?

An excerpt from my experiments:

```js
// Module to generate a UI from a prompt + initial state payload
export function GeneratedUI(id, prompt, localState) {
  const render$ = new Subject();
  const generate$ = new BehaviorSubject(prompt);
  const html$ = new BehaviorSubject("");

  // map over state and create a new BehaviorSubject for each key
  const state$ = Object.keys(localState).reduce((acc, key) => {
    acc[key] = new BehaviorSubject(localState[key]);
    return acc;
  }, {});

  const generatedHtml$ = generate$.pipe(imagine(id), tap(debug));

  const ui$ = render$.pipe(
    map(() => render(id, html$.getValue(), state(state$))),
  );

  connect(generatedHtml$, html$);
  connect(html$, render$);
  Object.keys(state$).forEach((key) => {
    connect(state$[key], render$);
  });

  return {
    in: {
      render: render$,
      generate: generate$,
    },
    out: {
      ui: ui$,
      html: html$,
      ...state$,
    },
  };
}
```

You could argue that the `html` is internal state here, as no-one cares about it except this components render pipeline, but it can still be exported as an output.

Reply:
to clarify, is this code that generates a module, or on its own a module? if the latter, where are `prompt` and `localState` coming from when invoked in a recipe?